vertical-pod-autoscaler/enhancements/3913-customized-recomender-vpa.md

```diff
@@ -0,0 +1,272 @@
+# Support Customized Recommenders for Vertical Pod Autoscalers
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Implementation Details](#implementation-details)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+- [Alternatives](#alternatives)
+- [Out of Scope](#out-out-scope)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+Today, the current VPA recommends CPU/Mem requests and limits based on one recommender, 
+which recommends the future requests/limits based on the historical usage observed in a 
+rolling time window. As there is no universal recommendation policy that applies to all 
+types of workload, this KEP suggests supporting multiple customized recommenders in VPA.  
+Thus, users can run different recommenders for different workloads, as they may exhibit 
+very distinct resource usage behaviors.
+
+## Motivation
+
+A VPA is used to resize containers in pods when the actual CPU/memory usage of a container 
+is significantly different from the resources requested. Resource usage-based recommendation 
+is the basic approach that resizes containers according to the actual usage observed and is 
+implemented in the default VPA recommender. Users can configure the time window and a certain 
+percentile of observed usage in the past as the prediction of future requests/limits for CPU/memory.
+
+However, as containers running different types of workloads may have different resource usage patterns, 
+there is no universal policy that applies to all. The existing VPA recommender may not accurately 
+predict future resource usage when containers exhibit certain resource usage behaviors, 
+such as trending, periodically changing, or occasional spikes, resulting in significant 
+over-provisioning and OOM kills for microservices. Learning different types of resource usage 
+behaviors for workloads and applying different algorithms to improve resource utilization 
+(CPU and Memory) predictions can significantly reduce over-provisioning and OOM kills in VPA.
+
+### Goals
+
+- Allow the VPA object to specify a customized recommender to use. 
+- Allow the VPA object to use the default recommender when no recommender is specified.
+
+### Non-Goals
+
+- We assume no pod uses two recommenders at the same time.
```

Comment:
I wonder how much we need this assumption.

Just to be clear: I'm not suggesting the we handle multiple recommenders already but if we could just make sure we don't close the way towards that goal w/o introducing much extra work, I'd be for it.

There are at least two use-cases which might require multiple recommenders:
* recommendation depending on pod life-cycle phase - admittedly this requires in-place updates to be fully useful
* separate recommenders for separate resources (e.g. you could have few dedicated recommender for CPU patterns, a separate set for memory patterns and let users mix them up according to their needs, w/o deploying all the combinations).